pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -0,0 +1,549 @@
+	name string,
+	informerIndex krt.Index[cluster.ID, krt.Collection[T]],
+	opts krt.OptionsBuilder,
+) krt.TransformationSingle[Cluster, krt.Collection[config.ObjectWithCluster[T]]] {
+	return func(ctx krt.HandlerContext, c Cluster) *krt.Collection[config.ObjectWithCluster[T]] {
+		clients := informerIndex.Lookup(c.ID)
+		if len(clients) == 0 {
+			return nil
+		}
+		objWithCluster := krt.NewCollection(clients[0], func(ctx krt.HandlerContext, obj T) *config.ObjectWithCluster[T] {
+			return &config.ObjectWithCluster[T]{ClusterID: c.ID, Object: &obj}
+		}, opts.With(
+			krt.WithName(name+"WithCluster"),
+			krt.WithMetadata(krt.Metadata{
+				ClusterKRTMetadataKey: c.ID,
+			}),
+		)...)
+		return &objWithCluster
+	}
+}
+
+func informerIndexByCluster[T controllers.ComparableObject](
+	informerCollection krt.Collection[krt.Collection[T]],
+) krt.Index[cluster.ID, krt.Collection[T]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[T]](informerCollection, func(col krt.Collection[T]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func nestedCollectionIndexByCluster[T any](
+	collection krt.Collection[krt.Collection[config.ObjectWithCluster[T]]],
+) krt.Index[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]](collection, func(col krt.Collection[config.ObjectWithCluster[T]]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func mergeServiceInfosWithCluster(localClusterID cluster.ID) func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+	return func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+		if len(serviceInfos) == 0 {
+			return nil
+		}
+		if len(serviceInfos) == 1 {
+			return &serviceInfos[0]
+		}
+
+		var vips []*workloadapi.NetworkAddress
+		var sans []string
+		var base config.ObjectWithCluster[model.ServiceInfo]
+		for _, obj := range serviceInfos {
+			if obj.Object == nil {
+				continue
+			}
+
+			vips = append(vips, obj.Object.Service.GetAddresses()...)
+			sans = append(sans, obj.Object.Service.GetSubjectAltNames()...)
+			if obj.ClusterID == localClusterID {
+				// This is the base object we'll append to since it's local
+				base = obj
+			}
+		}
+
+		// TODO: Do we need to merge anything else?
```

Comment:
In current logic we have a ton of semantics around precedence around Service vs ServiceEntry, merging ports, etc

Reply:
[Here](https://github.com/istio/istio/blob/29ea0d4b9398115dbfd6bea6019db28c91fc71a3/pilot/pkg/serviceregistry/aggregate/controller.go#L271) you mean? I'd forgotten to check this; thanks for pointing it out. Seems like I can encode this into the merge function